### PR TITLE
Implement the InputEvent

### PR DESCRIPTION
**Pull request by pyfisch**
Add WebIDL, from UI Events working draft.
Implement script type for event.

Note that actually firing InputEvents is not part of this commit.

As part of my effort to improve [keyboard handling](https://groups.google.com/forum/#!topic/mozilla.dev.servo/e6R7eBDjo4M) it is necessary to introduce the InputEvent. While I have a working prototype using the input event this still needs a lot of work  before it can be merged into servo.

cc @jdm

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17578)

### CODE REVIEW

File: components/script/dom/inputevent.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+use dom::uievent::UIEvent;
+use dom::window::Window;
+use dom_struct::dom_str;
+use std::cell::Cell;
+
+#[dom_struct]
+pub struct InputEvent {
+    uievent: UIEvent,
+    data: DOMRefCell<Option<DOMString>>,
+    is_composing: Cell<bool>,
```
Comment (jdm) [inline components/script/dom/inputevent.rs]: We can remove the DOMRefCell and Cell if we remove new_uninitialized and make new_inherited take the initial values.

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/dom/inputevent.rs (hunk 1 above)
Comment (pyfisch) [inline components/script/dom/inputevent.rs]: The current definition with DOMRefCell and Cell is similiar to [KeyboardEvent](https://github.com/servo/servo/blob/master/components/script/dom/keyboardevent.rs). Is there a reason they should be different?

Comment (jdm) [inline components/script/dom/inputevent.rs]: KeyboardEvent can be created with document.createEvent so it needs to support uninitialized default values. InputEvent cannot: https://dom.spec.whatwg.org/#dom-document-createevent

**Comment by pyfisch** [comment]
@jdm thanks for explaining. Removed Cell and DOMRefCell.

**Comment by jdm** [comment]
You will need to add the interface to https://dxr.mozilla.org/servo/source/tests/wpt/mozilla/tests/mozilla/interfaces.html#12 so the test does not fail.

**Comment by pyfisch** [comment]
Added the interface.

@bors-servo try

**Comment by pyfisch** [comment]
Travis had a LLVM panic with:
``` 
   Compiling style v0.0.1 (file:///home/travis/build/servo/servo/components/style)
LLVM ERROR: IO failure on output stream.
error: Could not compile `style`.

To learn more, run the command again with --verbose.

The command "./mach test-unit" exited with 101.
```
I guess this is an intermittent failure but I did not find an issue for it.

The failing wpt test is because `InputEventInit.data` has `""` as default value but according to the test it should be null. I have submitted a PR at w3c/uievents#148.

**Comment by pyfisch** [comment]
I have not yet changed the InputEventInit.data default value. The tests will fail again wit the same message.
Thought that my PR to the specification would get some review from the people at w3c and I planned to update this afterwards but I can also change it here.

**Comment by pyfisch** [comment]
@jdm Added null. r+?

**Pull request merged**